pinot-broker/src/main/java/com/linkedin/pinot/broker/queryquota/TableQueryQuotaManager.java

```diff
@@ -0,0 +1,267 @@
+    TableConfig tableConfig = ZKMetadataProvider.getTableConfig(propertyStore, tableNameWithType);
+    if (tableConfig == null) {
+      return null;
+    }
+    return tableConfig.getQuotaConfig();
+  }
+
+  /**
+   * Create a rate limiter for a table.
+   * @param tableName table name with table type.
+   * @param brokerResource broker resource which stores all the broker states of each table.
+   * @param quotaConfig quota config of the table.
+   * */
+  private void createRateLimiter(String tableName, ExternalView brokerResource, QuotaConfig quotaConfig) {
+    if (quotaConfig == null || Strings.isNullOrEmpty(quotaConfig.getMaxQueriesPerSecond())) {
+      LOGGER.info("No qps config specified for table: {}", tableName);
+      return;
+    }
+
+    if (brokerResource == null) {
+      LOGGER.warn("Failed to init qps quota for table {}. No broker resource connected!", tableName);
+      return;
+    }
+
+    Map<String, String> stateMap = brokerResource.getStateMap(tableName);
+    if (stateMap == null) {
+      LOGGER.warn("Failed to init qps quota: table {} is not in the broker resource!", tableName);
+      return;
+    }
+    int onlineCount = 0;
+    for (String state : stateMap.values()) {
+      if (state.equals(CommonConstants.Helix.StateModel.SegmentOnlineOfflineStateModel.ONLINE)) {
+        onlineCount++;
+      }
+    }
+    LOGGER.info("The number of online brokers for table {} is {}", tableName, onlineCount);
+
+    // assert onlineCount > 0;
+    if (onlineCount == 0) {
+      LOGGER.warn("Failed to init qps quota: there's no online broker services for table {}!", tableName);
+//      return;
+    }
+
+    // FIXME We use fixed rate for the 1st version.
+    onlineCount = 1;
+
+    // Get the dynamic rate
+    double overallRate;
+    if (quotaConfig.isMaxQueriesPerSecondValid()) {
+      overallRate = Double.parseDouble(quotaConfig.getMaxQueriesPerSecond());
+    } else {
+      LOGGER.error("Failed to init qps quota: error when parsing qps quota: {} for table: {}",
+          quotaConfig.getMaxQueriesPerSecond(), tableName);
+      return;
+    }
+
+    double perBrokerRate =  overallRate / onlineCount;
+    QueryQuotaConfig queryQuotaConfig = new QueryQuotaConfig(RateLimiter.create(perBrokerRate), new HitCounter(TIME_RANGE_IN_SECOND));
+    _rateLimiterMap.put(tableName, queryQuotaConfig);
+    LOGGER.info("Rate limiter for table: {} has been initialized. Overall rate: {}. Per-broker rate: {}. Number of online broker instances: {}",
+        tableName, overallRate, perBrokerRate, onlineCount);
+  }
+
+  /**
+   * Acquire a token from rate limiter based on the table name.
+   * @param tableName original table name which could be raw.
+   * @return true if there is no query quota specified for the table or a token can be acquired, otherwise return false.
+   * */
+  public boolean acquire(String tableName, String offlineTableName, String realtimeTableName) {
```

Comment:
we do we pass 3 table names in the argument? why not just pass the raw table name and construct the other two s needed?